File: apps/backend/models/recurring.py
from datetime import date, datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from models.enums import (
    AlertSeverity,
    AlertStatus,
    AlertType,
    FrequencyType,
    StreamStatus,
    StreamType,
    UserActionType,
)


class RecurringStreamBase(BaseModel):
    stream_type: StreamType
    description: str
    merchant_name: str | None = None
    category_primary: str | None = None
    category_detailed: str | None = None
    frequency: FrequencyType
    first_date: date
    last_date: date
    predicted_next_date: date | None = None
    average_amount: Decimal
    last_amount: Decimal
    iso_currency_code: str = "USD"
    is_active: bool = True
    status: StreamStatus
    is_user_modified: bool = False
    transaction_ids: list[str] = Field(default_factory=list)


class RecurringStreamCreate(RecurringStreamBase):
    user_id: UUID
    plaid_item_id: UUID
    account_id: UUID
    stream_id: str
    plaid_raw: dict[str, Any] | None = None


class RecurringStreamUpdate(BaseModel):
    description: str | None = None
    merchant_name: str | None = None
    category_primary: str | None = None
    category_detailed: str | None = None
    frequency: FrequencyType | None = None
    last_date: date | None = None
    predicted_next_date: date | None = None
    average_amount: Decimal | None = None
    last_amount: Decimal | None = None
    is_active: bool | None = None
    status: StreamStatus | None = None
    is_user_modified: bool | None = None
    transaction_ids: list[str] | None = None
    plaid_raw: dict[str, Any] | None = None
    last_synced_at: datetime | None = None


class RecurringStreamResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    plaid_item_id: UUID
    account_id: UUID
    stream_id: str
    stream_type: StreamType
    description: str
    merchant_name: str | None
    category_primary: str | None
    category_detailed: str | None
    frequency: FrequencyType
    first_date: date
    last_date: date
    predicted_next_date: date | None
    average_amount: Decimal
    last_amount: Decimal
    iso_currency_code: str
    is_active: bool
    status: StreamStatus
    is_user_modified: bool
    transaction_ids: list[str]
    last_synced_at: datetime
    created_at: datetime
    updated_at: datetime


class RecurringStreamListResponse(BaseModel):
    inflow_streams: list[RecurringStreamResponse]
    outflow_streams: list[RecurringStreamResponse]
    total_monthly_inflow: Decimal
    total_monthly_outflow: Decimal
    last_synced_at: datetime | None


class UpcomingBillResponse(BaseModel):
    stream: RecurringStreamResponse
    days_until_due: int
    expected_amount: Decimal


class UpcomingBillsListResponse(BaseModel):
    bills: list[UpcomingBillResponse]
    total_amount: Decimal
    period_days: int


class AlertBase(BaseModel):
    alert_type: AlertType
    severity: AlertSeverity
    title: str
    message: str
    data: dict[str, Any] | None = None


class AlertCreate(AlertBase):
    user_id: UUID
    recurring_stream_id: UUID | None = None


class AlertUpdate(BaseModel):
    status: AlertStatus | None = None
    user_action: UserActionType | None = None
    read_at: datetime | None = None
    dismissed_at: datetime | None = None
    actioned_at: datetime | None = None


class AlertResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    recurring_stream_id: UUID | None
    alert_type: AlertType
    severity: AlertSeverity
    title: str
    message: str
    data: dict[str, Any] | None
    status: AlertStatus
    user_action: UserActionType | None
    created_at: datetime
    read_at: datetime | None
    dismissed_at: datetime | None
    actioned_at: datetime | None


class AlertWithStreamResponse(AlertResponse):
    stream: RecurringStreamResponse | None = None


class AlertListResponse(BaseModel):
    alerts: list[AlertWithStreamResponse]
    total: int
    unread_count: int


class AlertAcknowledgeRequest(BaseModel):
    user_action: UserActionType | None = None


class RecurringSyncResult(BaseModel):
    streams_synced: int
    streams_created: int
    streams_updated: int
    streams_deactivated: int
    alerts_created: int


class StreamTransactionResponse(BaseModel):
    id: UUID
    account_id: UUID
    transaction_id: str
    amount: Decimal
    iso_currency_code: str
    date: date
    name: str
    merchant_name: str | None = None
    pending: bool
    payment_channel: str | None = None
    category: list[str] | None = None
    personal_finance_category_primary: str | None = None
    personal_finance_category_detailed: str | None = None
    logo_url: str | None = None
    created_at: datetime
    updated_at: datetime


class RecurringStreamDetailResponse(BaseModel):
    stream: RecurringStreamResponse
    transactions: list[StreamTransactionResponse]
    total_spent: Decimal
    occurrence_count: int

File: apps/mobile/types/recurring.ts
export type StreamType = 'inflow' | 'outflow';

export type FrequencyType =
  | 'weekly'
  | 'biweekly'
  | 'semi_monthly'
  | 'monthly'
  | 'quarterly'
  | 'semi_annually'
  | 'annually'
  | 'irregular'
  | 'unknown';

export type StreamStatus = 'mature' | 'early_detection' | 'tombstoned';

export type AlertType =
  | 'price_increase'
  | 'price_decrease'
  | 'new_subscription'
  | 'cancelled_subscription'
  | 'missed_payment';

export type AlertSeverity = 'low' | 'medium' | 'high';

export type AlertStatus = 'unread' | 'read' | 'dismissed' | 'actioned';

export type UserActionType = 'dismissed' | 'cancelled_subscription' | 'kept' | 'watching';

export interface RecurringStream {
  id: string;
  user_id: string;
  plaid_item_id: string;
  account_id: string;
  stream_id: string;
  stream_type: StreamType;
  description: string;
  merchant_name: string | null;
  category_primary: string | null;
  category_detailed: string | null;
  frequency: FrequencyType;
  first_date: string;
  last_date: string;
  predicted_next_date: string | null;
  average_amount: number;
  last_amount: number;
  iso_currency_code: string;
  is_active: boolean;
  status: StreamStatus;
  is_user_modified: boolean;
  transaction_ids: string[];
  last_synced_at: string;
  created_at: string;
  updated_at: string;
}

export interface RecurringStreamListResponse {
  inflow_streams: RecurringStream[];
  outflow_streams: RecurringStream[];
  total_monthly_inflow: number;
  total_monthly_outflow: number;
  last_synced_at: string | null;
}

export interface UpcomingBill {
  stream: RecurringStream;
  days_until_due: number;
  expected_amount: number;
}

export interface UpcomingBillsListResponse {
  bills: UpcomingBill[];
  total_amount: number;
  period_days: number;
}

export interface RecurringSyncResult {
  streams_synced: number;
  streams_created: number;
  streams_updated: number;
  streams_deactivated: number;
  alerts_created: number;
}

export interface Alert {
  id: string;
  user_id: string;
  recurring_stream_id: string | null;
  alert_type: AlertType;
  severity: AlertSeverity;
  title: string;
  message: string;
  data: Record<string, string | number | null> | null;
  status: AlertStatus;
  user_action: UserActionType | null;
  created_at: string;
  read_at: string | null;
  dismissed_at: string | null;
  actioned_at: string | null;
}

export interface AlertWithStream extends Alert {
  stream: RecurringStream | null;
}

export interface AlertListResponse {
  alerts: AlertWithStream[];
  total: number;
  unread_count: number;
}

export interface AlertAcknowledgeRequest {
  user_action?: UserActionType;
}

export interface StreamTransaction {
  id: string;
  account_id: string;
  transaction_id: string;
  amount: number;
  iso_currency_code: string;
  date: string;
  name: string;
  merchant_name: string | null;
  pending: boolean;
  payment_channel: string | null;
  category: string[] | null;
  personal_finance_category_primary: string | null;
  personal_finance_category_detailed: string | null;
  logo_url: string | null;
  created_at: string;
  updated_at: string;
}

export interface RecurringStreamDetailResponse {
  stream: RecurringStream;
  transactions: StreamTransaction[];
  total_spent: number;
  occurrence_count: number;
}
